### PR TITLE
Update Dig error types & reporting, more easily distinguish Dig errors vs. non-Dig errors

### PR DESCRIPTION
**Pull request by JacobOaks**
This update:
- Introduces the `dig.Error` interface, which is implemented by all errors returned by Dig, allowing users to use `errors.Is` and `errors.As` with Dig errors.
- Replaces the concept of errors causing other errors with Go's error wrapping conventions, so errors caused by other errors can now be unwrapped to their causing error with `errors.Unwrap`.
- Updates the behavior of `dig.RootCause` to return the first error non-`dig.Error` in the chain of errors passed to it, if there is one. Otherwise, it will return the first error that does not wrap another error (the bottom of the chain). In response to [this issue](https://github.com/uber-go/dig/issues/356), users can now do something like this:
```go
rootCause := dig.RootCause(err)
var de dig.Error
if errors.As(rootCause, &de) {
  // is a Dig error
} else {
  // is an error thrown by one of my provided or invoked functions
}
```
- Adds tests testing the new behavior of `dig.RootCause` in `error_test.go`

### CODE REVIEW

File: result.go [hunk 1]
```diff
@@ -70,35 +69,33 @@ type resultOptions struct {
...
+			return nil, errValueGroup{
+				fmt.Sprintf("attempted to use soft with result value group: %q", g.Name)}
 		}
 		if g.Flatten {
 			if t.Kind() != reflect.Slice {
-				return nil, errf(
-					"flatten can be applied to slices only",
-					"%v is not a slice", t)
+				return nil, newErrSpecification(
+					fmt.Sprintf("attempted to use flatten on non-slice: %v", t), nil)
```
Comment (sywhang) [inline result.go:98]: let's try to minimize updating error strings unless there's a good reason to.

File: result.go [hunk 2]
```diff
@@ -340,13 +339,13 @@ func (ro resultObject) DotResult() []*dot.Result {
...
-			"cannot specify a name for result objects", "%v embeds dig.Out", t)
+		return ro, newErrSpecification(
+			fmt.Sprintf("attempted to specify a name for result object which embeds dig.Out: %v", t), nil)
 	}
 
 	if len(opts.Group) > 0 {
-		return ro, errf(
-			"cannot specify a group for result objects", "%v embeds dig.Out", t)
+		return ro, newErrSpecification(
+			fmt.Sprintf("attempted to specify a group for result object which embeds dig.Out: %v", t), nil)
```
Comment (sywhang) [inline result.go:348]: same thing here.

File: visualize.go [hunk 3]
```diff
@@ -62,27 +63,28 @@ func (o visualizeErrorOption) applyVisualizeOption(opt *visualizeOptions) {
...
-			errors = append(errors, ev)
+			errs = append(errs, ev)
 		}
-		e, ok := err.(causer)
-		if !ok {
+		e := errors.Unwrap(err)
+		if e == nil {
 			break
+		} else {
+			err = e
```
Comment (sywhang) [inline visualize.go:76]: with `break`, you don't need this else block.

https://github.com/uber-go/guide/blob/master/style.md#reduce-nesting

File: visualize.go [hunk 4]
```diff
@@ -156,11 +158,12 @@ func CanVisualizeError(err error) bool {
 		if _, ok := err.(errVisualizer); ok {
 			return true
 		}
-		e, ok := err.(causer)
-		if !ok {
+		e := errors.Unwrap(err)
+		if e == nil {
 			break
+		} else {
+			err = e
```
Comment (sywhang) [inline visualize.go:165]: same issue here (nesting).

File: error.go [hunk 5]
```diff
@@ -27,187 +27,99 @@ import (
...
-					len(args))
-			}
+// wrappedError is a dig.Error with additional functionality
+// that allows them to be formatted via formatError()
+type wrappedError interface {
+	Error
+	fmt.Formatter
+	Unwrap() error
+	writeMessage(w io.Writer, v string)
+}
```
Comment (sywhang) [inline error.go:78]: I haven't gone through and verified every single dig.Error implementations to see if they implement wrappedError, but it seems like they do. In that case, can you just merge the definitions into one, and remove the dummy() method on dig.Error interface?

**Review by sywhang — COMMENTED** [review]
took a quick pass, will come back to it later tonight.

**Comment by JacobOaks** [comment]
I have pushed a new commit:
- combine `dig.Error` and `wrappedError` interfaces, remove `dummy()` methods
- rename `errSpecification` to `errInvalidInput`. I wanted to be careful to disambiguate the idea of user input error from errors specifically related to building arguments for functions in the Dig container (`errArgumentsFailed`), hence my avoiding using the term "argument".
- clean up code, bring up test coverage, address @sywhang's other points

File: error.go [hunk 6]
```diff
@@ -27,187 +27,99 @@ import (
...
+	Message string
+	Cause   error
+}
 
-	// Writes the message or context for this error in the chain.
-	//
-	// verb is either %v or %+v.
-	writeMessage(w io.Writer, verb string)
+// newErrSpecification creates a new errSpecification. If there is no other
+// error to wrap, it is safe to pass nil.
```
[the remark this answers is not in the document]
Comment (JacobOaks) [reply error.go]: I updated the comment which should explain what I meant by this

**Review by josephinedotlee — COMMENTED** [review]
oh behalf of @abhinav for his last internal code review 😭

**Comment by JacobOaks** [comment]
Update:
* RootCause() now more closely matches previous behavior: it will return first non-`dig.Error` or first error that does not wrap another error.
* There are now two interfaces for Dig errors. dig.Errors represents the public contract, and digError is an extension for private functionality (primarily error formatting). dig.Errors remains sealed.
* Add fields to structs in test tables

**Review by sywhang — APPROVED** [review]

**Pull request merged**